samza-core/src/main/java/org/apache/samza/control/WatermarkManager.java

```diff
@@ -0,0 +1,191 @@
+   *   <li>Watermark(stream) = min { Watermark(ssp) | ssp is a partition of stream that assigns to this task } </li>
+   * </ul>
+   *
+   * @param envelope the envelope contains {@link WatermarkMessage}
+   * @return watermark envelope if there is a new aggregate watermark for the stream
+   */
+  public Watermark update(IncomingMessageEnvelope envelope) {
+    SystemStreamPartition ssp = envelope.getSystemStreamPartition();
+    WatermarkState state = watermarkStates.get(ssp);
+    WatermarkMessage message = (WatermarkMessage) envelope.getMessage();
+    state.update(message.getTimestamp(), message.getTaskName(), message.getTaskCount());
+
+    if (state.getWatermarkTime() != TIME_NOT_EXIST) {
+      long minTimestamp = watermarkStates.entrySet().stream()
+          .filter(entry -> entry.getKey().getSystemStream().equals(ssp.getSystemStream()))
+          .map(entry -> entry.getValue().getWatermarkTime())
+          .min(Long::compare)
+          .get();
+      Long curWatermark = watermarkPerStream.get(ssp.getSystemStream());
+      if (curWatermark == null || curWatermark < minTimestamp) {
+        watermarkPerStream.put(ssp.getSystemStream(), minTimestamp);
+        return new WatermarkImpl(minTimestamp);
+      }
+    }
+
+    return null;
+  }
+
+  /* package private */
+  long getWatermarkTime(SystemStreamPartition ssp) {
+    return watermarkStates.get(ssp).getWatermarkTime();
+  }
+
+  /**
+   * Send the watermark message to all partitions of an intermediate stream
+   * @param timestamp watermark timestamp
+   * @param systemStream intermediate stream
+   */
+  void sendWatermark(long timestamp, SystemStream systemStream, int taskCount) {
+    log.info("Send end-of-stream messages to all partitions of " + systemStream);
+    final WatermarkMessage watermarkMessage = new WatermarkMessage(timestamp, taskName, taskCount);
+    ControlMessageUtils.sendControlMessage(watermarkMessage, systemStream, metadataCache, collector);
+  }
+
+  /**
+   * Per ssp state of the watermarks. This class keeps track of the latest watermark timestamp
+   * from each upstream producer tasks, and use the min to update the aggregated watermark time.
+   */
+  final static class WatermarkState {
+    private int expectedTotal = Integer.MAX_VALUE;
+    private final Map<String, Long> timestamps = new HashMap<>();
+    private long watermarkTime = TIME_NOT_EXIST;
+
+    void update(long timestamp, String taskName, int taskCount) {
+      if (taskName != null) {
+        timestamps.put(taskName, timestamp);
+      }
+      expectedTotal = taskCount;
+
+      if (timestamps.size() == expectedTotal) {
+        Optional<Long> min = timestamps.values().stream().min(Long::compare);
+        watermarkTime = min.orElse(timestamp);
+      }
+    }
+
+    long getWatermarkTime() {
+      return watermarkTime;
+    }
+  }
+
+  /**
+   * Implementation of the Watermark. It keeps a reference to the {@link WatermarkManager}
```

Comment:
We seem to be mixing two ideas here... the watermark itself with its mechanism for propagation. It would be cleaner to provide these at the fluent API level (this is how flink works) and to do the right default operation for typical operations. Ultimately the user needs to be able to override this behavior though and I'm not seeing that capability so far in the design.